### ISSUE TITLE
Allow for running jobs in sequence

### ISSUE THREAD

**Issue by zachahn** [zachahn/disqualified#4]
I would like an API that lets me run jobs in sequence. This feature would only work with Disqualified jobs, not through the ActiveJob integration.

Just some ideas:

```ruby
# I don't like this API because it requires a "run" at the end. This is easy to forget.
# Also, how do we handle `perform_in`? Should `MyNextJob` run 5 minutes after queuing, or 5 minutes after `MyFirstJob` finishes?
Disqualified::Sequence
  .new { MyFirstJob.perform_async("test") }
  .then { MyNextJob.perform_in(5.minutes, "test") }
  .then { MyFinalJob.perform_async("test") }
  .run

# Any API that uses a kind of factory pattern REQUIRES an explicit "build" method ("run" in this case).
# (I don't like this API)
Disqualified::Sequence
  .new(MyFirstJob, "test")
  .then_in(5.minutes, MyNextJob, "test")
  .then(MyFinalJob, "test")
  .run

# Using a block is a really convenient way of knowing when the job is ready to run
Disqualified::Sequence.new do |s|
  s.run(MyFirstJob, "test")
  s.run_in(5.minutes, MyNextJob, "test")
  s.run(MyFinalJob, "test")
end

# I like that this requires very little new API.
# However, all 3 jobs here would be enqueued at the same time. (They will just be run in order after the previous is done.)
# Perhaps I should raise an error if `perform_in` gets called inside a sequence block.
# And actually... I might be able to get this to work with the ActiveJob integration
# Note that none of these APIs (this one and above) can have "on_error" callbacks because I don't have a way of storing the callback code / state anywhere.
Disqualified::Sequence.queue do
  MyFirstJob.perform_async("test")
  MyNextJob.perform_in(5.minutes, "test")
  MyFinalJob.perform_async("test")
end

# This is a bit verbose, but it's nice that it's a normal class definition
class MySequentialJob
  include Disqualified::Job
  include Disqualified::Sequence

  def perform(args)
    sequence do
      MyFirstJob.perform_async("test")
      MyNextJob.perform_in(5.minutes, "test")
      MyFinalJob.perform_async("test")
    end
  end

  def on_sequence_error(e, metadata)
  end
end

# Note that the previous previous API would still allow for a class definition
class MySequentialJob
  include Disqualified::Job

  def perform(arg)
    Disqualified::Sequence.queue do
      MyFirstJob.perform_async(arg)
      MyNextJob.perform_in(5.minutes, arg)
      MyFinalJob.perform_async(arg)
    end
  end
end
```

**Comment by zachahn** [comment]
I settled on one of the approaches above—reiterating it here for posterity:

```ruby
# Note that none of these APIs (this one and above) can have "on_error" callbacks because I don't have a way of storing the callback code / state anywhere.
Disqualified::Sequence.queue do
  MyFirstJob.perform_async("test")
  MyNextJob.perform_in(5.minutes, "test")
  MyFinalJob.perform_async("test")
end
```

A few notes:

- This works with any combination of ActiveJob and Disqualified jobs (as long as ActiveJob's queue adapter is Disqualified)
- There is no special handling for `#perform_in` / `#perform_at`. All jobs in a sequence will be queued immediately, but only run once it's its turn to run AND it's time for it to run
- There is no "batching" feature here, just a simple sequencing feature. Jobs queued within a sequence will run outside the context of a sequence; there is no way to mutate an existing sequence

This lives in my `sequence` branch. I'm testing it in one of my apps, and I'm happy with it so far. I'm likely to merge this in and release very soon.

**Comment by zachahn** [comment]
Released in v0.5.0
